json-rpc/src/stream_rpc/subscriptions.rs

```diff
@@ -0,0 +1,258 @@
+                        helper.method.as_str(),
+                        counters::RpcResult::Success.as_str(),
+                        helper.client.connection_context.sdk_info.language.as_str(),
+                        &helper
+                            .client
+                            .connection_context
+                            .sdk_info
+                            .version
+                            .to_string(),
+                    ])
+                    .inc_by(items.len() as u64);
+
+                for item in items {
+                    match helper
+                        .client
+                        .send_success(helper.jsonrpc_id.clone(), &item)
+                        .await
+                    {
+                        Err(e) => {
+                            // client has disconnected
+                            debug!("Client#{}: Send error: {:?}", &helper.client.id, e);
+                            return;
+                        }
+                        Ok(_) => self.on_send(&item),
+                    };
+                }
+            }
+        }))
+    }
+}
+
+impl Subscription<TransactionView> for SubscribeToTransactionsParams {
+    fn init(&mut self, _helper: &SubscriptionHelper) -> Result<(), JsonRpcError> {
+        self.latest_version = self.starting_version;
+        Ok(())
+    }
+
+    fn next(&self, helper: &SubscriptionHelper) -> Vec<TransactionView> {
+        match get_transactions(
+            helper.db.borrow(),
+            helper.db.get_latest_version().unwrap_or(0),
+            self.latest_version,
+            helper.client.config.fetch_size,
+            self.include_events.unwrap_or(false),
+        ) {
+            Ok(transactions) => transactions.0,
+            Err(e) => {
+                warn!(
+                    "Client#{} Could not fetch transactions: {}",
+                    helper.client.id, e
+                );
+                vec![]
+            }
+        }
+    }
+
+    fn on_send(&mut self, _tx: &TransactionView) {
+        self.latest_version += 1;
+    }
+}
+
+impl Subscription<EventView> for SubscribeToEventsParams {
+    fn init(&mut self, _helper: &SubscriptionHelper) -> Result<(), JsonRpcError> {
+        self.latest_event = self.event_seq_num;
+        Ok(())
+    }
+
+    fn next(&self, helper: &SubscriptionHelper) -> Vec<EventView> {
+        match get_events(
+            helper.db.borrow(),
+            helper.db.get_latest_version().unwrap_or(0),
+            self.event_key,
+            self.latest_event,
+            helper.client.config.fetch_size,
+        ) {
+            Ok(events) => events,
+            Err(e) => {
+                warn!("Client#{} Could not fetch events: {}", helper.client.id, e);
+                vec![]
+            }
+        }
+    }
+
+    fn on_send(&mut self, event: &EventView) {
+        self.latest_event = event.sequence_number + 1
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn test() {}
```

Comment:
This isn't the easiest sort of code to test, but I think just one happy path test would help drive out your architecture/methods to be more testable. ie How do you inject test input and assert on the output, esp with dependencies like ClientConnection?

I know it's a draft so will wait.

Reply:
How's this?